jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -906,12 +914,16 @@ private void releaseChunk()
         @Override
         protected void onCompleteSuccess()
         {
+            if (_shutdownOut && !(_httpChannel.getRequest().getAttribute(HttpStream.UPGRADE_CONNECTION_ATTRIBUTE) instanceof Connection))
+                getEndPoint().shutdownOutput();
             release().succeeded();
         }
 
         @Override
         public void onCompleteFailure(final Throwable x)
         {
+            if (_shutdownOut)
+                getEndPoint().shutdownOutput();
```

Comment:
I think this should be moved to `HttpStreamOverHTTP1.failed()`.

Here it seems way too early, as a failure here may still allow to write an error response.

Reply:
Please read the issue and comments.   
This is just reverting to how we have shutdown for decades.  I see no reason why we started deferring the shutdown in 12.0 so that it could be delayed by application code long after the response is committed.

Reply:
Yet, in case of failures, it may be possible to fail the write and still wanting to send an error response instead, for example for headers too large where the response is not committed as we fail before, and failing the channel callback would allow to send a 431, but not if we shutdown the endpoint here.